### ISSUE TITLE
ISE  'NoneType' object has no attribute 'is_single'

### ISSUE THREAD

**Issue by zchenb** [edgedb/edgedb#5211]
EdgeDB 3.0-dev.7414+local (repl 3.0.0-dev)

Query:
```
with n0 := (insert Note {name := "boxing", note := {}}),
     n1 := (insert Note {name := "unboxing", note := "lolol"}),
insert Person {name := "Phil Emarg", notes :=  (select Note {@metanote := "arg!"})};
```

Error:
```
edgedb error: InternalServerError: 'NoneType' object has no attribute 'is_single'
  Hint: This is most likely a bug in EdgeDB. Please consider opening an issue ticket at https://github.com/edgedb/edgedb/issues/new?template=bug_report.md
  Server traceback:
      Traceback (most recent call last):
        File "/Users/edgedbzb/work/edgedb/edb/pgsql/compiler/__init__.py", line 134, in compile_ir_to_sql_tree
          qtree = dispatch.compile(ir_expr, ctx=ctx)
        File "/usr/local/Cellar/python@3.10/3.10.10_1/Frameworks/Python.framework/Versions/3.10/lib/python3.10/functools.py", line 889, in wrapper
          return dispatch(args[0].__class__)(*args, **kw)
        File "/Users/edgedbzb/work/edgedb/edb/pgsql/compiler/expr.py", line 60, in compile_Set
          _compile_set_impl(ir_set, ctx=ctx)
        File "/Users/edgedbzb/work/edgedb/edb/pgsql/compiler/expr.py", line 114, in _compile_set_impl
          _compile_set(ir_set, ctx=ctx)
        File "/Users/edgedbzb/work/edgedb/edb/pgsql/compiler/expr.py", line 680, in _compile_set
          relgen.get_set_rvar(ir_set, ctx=ctx)
        File "/Users/edgedbzb/work/edgedb/edb/pgsql/compiler/relgen.py", line 164, in get_set_rvar
          return _process_toplevel_query(ir_set, ctx=ctx)
        File "/Users/edgedbzb/work/edgedb/edb/pgsql/compiler/relgen.py", line 287, in _process_toplevel_query
          rvars = _get_set_rvar(ir_set, ctx=ctx)
        File "/Users/edgedbzb/work/edgedb/edb/pgsql/compiler/relgen.py", line 350, in _get_set_rvar
          return process_set_as_subquery(ir_set, ctx=ctx)
        File "/Users/edgedbzb/work/edgedb/edb/pgsql/compiler/relgen.py", line 1385, in process_set_as_subquery
          dispatch.visit(expr, ctx=newctx)
        File "/usr/local/Cellar/python@3.10/3.10.10_1/Frameworks/Python.framework/Versions/3.10/lib/python3.10/functools.py", line 889, in wrapper
          return dispatch(args[0].__class__)(*args, **kw)
        File "/Users/edgedbzb/work/edgedb/edb/pgsql/compiler/dispatch.py", line 44, in visit
          compile(ir, ctx=ctx)
        File "/usr/local/Cellar/python@3.10/3.10.10_1/Frameworks/Python.framework/Versions/3.10/lib/python3.10/functools.py", line 889, in wrapper
          return dispatch(args[0].__class__)(*args, **kw)
        File "/Users/edgedbzb/work/edgedb/edb/pgsql/compiler/stmt.py", line 150, in compile_InsertStmt
          parts = dml.init_dml_stmt(stmt, parent_ctx=parent_ctx, ctx=ctx)
        File "/Users/edgedbzb/work/edgedb/edb/pgsql/compiler/dml.py", line 96, in init_dml_stmt
          clauses.compile_dml_bindings(ir_stmt, ctx=ctx)
        File "/Users/edgedbzb/work/edgedb/edb/pgsql/compiler/clauses.py", line 253, in compile_dml_bindings
          dispatch.compile(binding, ctx=bctx)
        File "/usr/local/Cellar/python@3.10/3.10.10_1/Frameworks/Python.framework/Versions/3.10/lib/python3.10/functools.py", line 889, in wrapper
          return dispatch(args[0].__class__)(*args, **kw)
        File "/Users/edgedbzb/work/edgedb/edb/pgsql/compiler/expr.py", line 60, in compile_Set
          _compile_set_impl(ir_set, ctx=ctx)
        File "/Users/edgedbzb/work/edgedb/edb/pgsql/compiler/expr.py", line 110, in _compile_set_impl
          _compile_set(ir_set, ctx=scopectx)
        File "/Users/edgedbzb/work/edgedb/edb/pgsql/compiler/expr.py", line 680, in _compile_set
          relgen.get_set_rvar(ir_set, ctx=ctx)
        File "/Users/edgedbzb/work/edgedb/edb/pgsql/compiler/relgen.py", line 222, in get_set_rvar
          rvars = _get_set_rvar(ir_set, ctx=subctx)
        File "/Users/edgedbzb/work/edgedb/edb/pgsql/compiler/relgen.py", line 350, in _get_set_rvar
          return process_set_as_subquery(ir_set, ctx=ctx)
        File "/Users/edgedbzb/work/edgedb/edb/pgsql/compiler/relgen.py", line 1385, in process_set_as_subquery
          dispatch.visit(expr, ctx=newctx)
        File "/usr/local/Cellar/python@3.10/3.10.10_1/Frameworks/Python.framework/Versions/3.10/lib/python3.10/functools.py", line 889, in wrapper
          return dispatch(args[0].__class__)(*args, **kw)
        File "/Users/edgedbzb/work/edgedb/edb/pgsql/compiler/dispatch.py", line 44, in visit
          compile(ir, ctx=ctx)
        File "/usr/local/Cellar/python@3.10/3.10.10_1/Frameworks/Python.framework/Versions/3.10/lib/python3.10/functools.py", line 889, in wrapper
          return dispatch(args[0].__class__)(*args, **kw)
        File "/Users/edgedbzb/work/edgedb/edb/pgsql/compiler/stmt.py", line 166, in compile_InsertStmt
          return dml.fini_dml_stmt(
        File "/Users/edgedbzb/work/edgedb/edb/pgsql/compiler/dml.py", line 461, in fini_dml_stmt
          clauses.compile_output(ir_stmt.result, ctx=ctx)
        File "/Users/edgedbzb/work/edgedb/edb/pgsql/compiler/clauses.py", line 227, in compile_output
          dispatch.visit(ir_set, ctx=newctx)
        File "/usr/local/Cellar/python@3.10/3.10.10_1/Frameworks/Python.framework/Versions/3.10/lib/python3.10/functools.py", line 889, in wrapper
          return dispatch(args[0].__class__)(*args, **kw)
        File "/Users/edgedbzb/work/edgedb/edb/pgsql/compiler/expr.py", line 85, in visit_Set
          _compile_set_impl(ir_set, ctx=ctx)
        File "/Users/edgedbzb/work/edgedb/edb/pgsql/compiler/expr.py", line 114, in _compile_set_impl
          _compile_set(ir_set, ctx=ctx)
        File "/Users/edgedbzb/work/edgedb/edb/pgsql/compiler/expr.py", line 684, in _compile_set
          _compile_shape(ir_set, ir_set.shape, ctx=ctx)
        File "/Users/edgedbzb/work/edgedb/edb/pgsql/compiler/expr.py", line 703, in _compile_shape
          result = shapecomp.compile_shape(ir_set, shape, ctx=ctx)
        File "/Users/edgedbzb/work/edgedb/edb/pgsql/compiler/shapecomp.py", line 89, in compile_shape
          card.is_single() and (
      AttributeError: 'NoneType' object has no attribute 'is_single'
      
      The above exception was the direct cause of the following exception:
      
      Traceback (most recent call last):
        File "/Users/edgedbzb/work/edgedb/edb/server/compiler_pool/worker_proc.py", line 55, in worker
          res = meth(*args)
        File "/Users/edgedbzb/work/edgedb/edb/server/compiler_pool/worker.py", line 161, in compile
          units, cstate = COMPILER.compile(
        File "/Users/edgedbzb/work/edgedb/edb/server/compiler/compiler.py", line 742, in compile
          unit_group = compile(ctx=ctx, source=source)
        File "/Users/edgedbzb/work/edgedb/edb/server/compiler/compiler.py", line 1842, in compile
          return _try_compile(ctx=ctx, source=source)
        File "/Users/edgedbzb/work/edgedb/edb/server/compiler/compiler.py", line 1910, in _try_compile
          comp, capabilities = _compile_dispatch_ql(
        File "/Users/edgedbzb/work/edgedb/edb/server/compiler/compiler.py", line 1816, in _compile_dispatch_ql
          query = _compile_ql_query(ctx, ql, script_info=script_info)
        File "/Users/edgedbzb/work/edgedb/edb/server/compiler/compiler.py", line 1394, in _compile_ql_query
          qtree, sql_text, argmap = pg_compiler.compile_ir_to_tree_and_sql(
        File "/Users/edgedbzb/work/edgedb/edb/pgsql/compiler/__init__.py", line 169, in compile_ir_to_tree_and_sql
          qtree, _ = compile_ir_to_sql_tree(
        File "/Users/edgedbzb/work/edgedb/edb/pgsql/compiler/__init__.py", line 151, in compile_ir_to_sql_tree
          raise errors.InternalServerError(*args) from e
      edb.errors.InternalServerError: 'NoneType' object has no attribute 'is_single'
edgedb> 
```

Schema:
```
module default {
type Note {
    required single property name -> str;
    optional single property note -> str;
}
type Person {
    required single property name -> str;
    optional multi property multi_prop -> str;
    multi link notes -> Note {
        property metanote -> str;
    }
    optional single property tag -> str;
}
type Foo {
    required single property val -> str;
    optional single property opt -> int64;
}
type Award {
    name : str;
}
type Card {
    single name : str;
    multi awards : Award;
    element : str;
    cost : int64;
}


type User {
    required name: str {
        delegated constraint exclusive;
    }

    multi deck: Card {
        count: int64 {
            default := 1;
        };
        property total_cost := @count * .cost;
    }

    property deck_cost := sum(.deck.cost);

    multi friends: User {
        nickname: str;
        # how the friend responded to requests for a favor
        #favor: array<bool>
    }

    multi awards: Award {
        constraint exclusive;
    }

    avatar: Card {
        text: str;
        property tag := .name ++ (("-" ++ @text) ?? "");
    }
}

}
```

**Comment by zchenb** [comment]
The suggested reason is that  with we don't associate a cardinality with binders.

**Comment by zchenb** [comment]
Surprisingly, this has been fixed on the master! Maybe as a part of #5508.
